### PR TITLE
Add `BaseDistribution.single` methods.

### PR DESCRIPTION
**Pull request by sile**
This PR adds `BaseDistribution.single` method.
This method is useful when implementing custom samplers.
After this PR is merged, I will use the method in [#399/skopt.py#L90](https://github.com/pfnet/optuna/pull/399/files#diff-665e262286e69f900727d7370b255753R90).

### CODE REVIEW

File: optuna/distributions.py [hunk 1]
```diff
@@ -48,6 +48,36 @@ def to_internal_repr(self, param_value_in_external_repr):
 
         return param_value_in_external_repr
 
+    @abc.abstractmethod
+    def empty(self):
+        # type: () -> bool
+        """Test whether the range of this distribution is empty.
+
+        When this method returns :obj:`True`, samplers cannot sample any value
```
Comment (toshihikoyanase) [inline optuna/distributions.py]: Could you add a link to `optuna.samplers`?

File: optuna/distributions.py [hunk 2]
```diff
@@ -48,6 +48,36 @@ def to_internal_repr(self, param_value_in_external_repr):
...
+        """
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def single(self):
+        # type: () -> bool
+        """Test whether the range of this distribution contains just a single value.
+
+        When this method returns :obj:`True`, samplers will always sample
```
Comment (toshihikoyanase) [inline optuna/distributions.py]: ditto.

File: optuna/distributions.py (hunk 1 above)
Comment (sile) [reply optuna/distributions.py]: Added: https://github.com/pfnet/optuna/pull/431/commits/64d9ca4d0008f8891a0fc48d2c8aa3b20aeb1dbe

File: optuna/distributions.py [hunk 3]
```diff
@@ -48,6 +48,36 @@ def to_internal_repr(self, param_value_in_external_repr):
 
         return param_value_in_external_repr
 
+    @abc.abstractmethod
+    def empty(self):
+        # type: () -> bool
+        """Test whether the range of this distribution is empty.
+
+        When this method returns :obj:`True`, :mod:`~optuna.samplers` cannot sample any value
+        from the distribution and raise an error when sampling.
```
Comment (g-votte) [inline optuna/distributions.py]: Please replace `raise` with `raises`.

**Pull request merged**